### PR TITLE
InputNode and tests

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

Delegated Input classes will be owned by their managing input class. For purposes of access they will be accessed through a vector of variant<refwrapper<InputClassA>, refwrapper<InputClassB>,... in the managing class. Since input classes can potentially be quite large it seems better to separate ownership and access. For that reason the ownership of the input class is through a vector of UniquePointers to InputNode a mixin interface that provides a virtual destructor for the input classes.

## What type(s) of changes does this code introduce?
- New feature
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Leconte 

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

File: src/io/InputNode.hpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace qmcplusplus
+{
+
+/** Interface to a class to be type erased for the purposes of ownership
+ *  To allow a vector of unique_ptr to InputNode to contain a heterolist of dynamic objects
+ */
+struct InputNode
+{
+  virtual ~InputNode(){};
```
Comment (ye-luo) [inline src/io/InputNode.hpp:25]: `= default` is better than `{}`

File: src/io/InputNode.hpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  template<typename T, typename VOWN, typename VINPUT, typename... Args>
+  static void append(VOWN& vown, VINPUT& vinput, Args&&... args)
```
Comment (ye-luo) [inline src/io/InputNode.hpp:27]: It doesn't seem to be related to `InputNode` at all.
Should vown be of type `std::vector<std::unique_ptr<InputNode>>` explicitly?

File: src/type_traits/template_types.hpp [hunk 3]
```diff
@@ -28,6 +28,10 @@ namespace qmcplusplus
  *  see: UPtrVector
  *  @{
  */
+
+template<typename T>
+using RefW = std::reference_wrapper<T>;
```
Comment (ye-luo) [inline src/type_traits/template_types.hpp:33]: `RefWrapper` is the shortest I can take. `RefW` is so confusing.

File: src/io/tests/test_InputNode.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  std::string name{"inputB"};
+  short sA{1};
+};
+
+using ChildInput = std::variant<RefW<InputA>, RefW<InputB>>;
+
+TEST_CASE("InputNode_BasicUse", "[io]")
+{
+  UPtrVector<InputNode> input_ownership;
+  std::vector<ChildInput> inputs;
```
Comment (ye-luo) [inline src/io/tests/test_InputNode.cpp:48]: I don't feel good to keep both input_ownership and inputs around.
Also don't see `variant` brings simplification.

```
enum class {
   DENISTY,
   ESTB
} EstimatorType;

template<typename ENUM>
InputNode
{
   ENUM type;
};

UPtrVector<InputNode<EstimatorType>> input_ownership;
```

In this way, we know the type with some restrictions also. Only need input_ownership.
Also header file only need to see the enum class instead of all the class definition.

Comment (PDoakORNL) [reply src/io/tests/test_InputNode.cpp:48]: Header file for EstimatorManagerInput doesn't see all the definitions the way I wrote it forward declaration is fine.  I can make this just use variant, the memory waste isn't probably worth worrying about.

Comment (ye-luo) [reply src/io/tests/test_InputNode.cpp:48]: No memory concern. My concern was that keeping tracking of two vector separately is error-prone

Comment (PDoakORNL) [reply src/io/tests/test_InputNode.cpp:48]: I think that is legitimate and going back and dropping the separate storage works well.  This was a premature optimization.

**Comment by PDoakORNL** [comment]
Input node likely premature memory optimization
